### PR TITLE
task/david-flek

### PR DESCRIPTION
**Pull request by Kylyi**

### CODE REVIEW

File: components/products/Product.vue [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+defineProps<{
+  product: Product;
+}>()
+
+</script>
+
+<template>
+  <div class="product">
+
+    <img v-if="product.image" :src="product.image" alt="Product Image" class="image" />
```
Comment (Kylyi) [inline components/products/Product.vue:13]: The image is missing some important properties, do you know which ones?

Also, I don't really like the behavior of the images when I resize the screen - the image stretches and malforms.
How would you fix that?

File: components/products/Product.vue [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    <div v-else class="image-placeholder"></div>
```
Comment (Kylyi) [inline components/products/Product.vue:14]: Even though I like the fact that you thought about putting an image placeholder here, I think there is a better approach to handle it.

For example, what would happen if the image "exists" (so there is the url coming from the backend) but the image is actually non-existent (was deleted for some reason)?

File: components/products/Product.vue [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    <div class="product-info">
+      <h1>{{ product.title }}</h1>
+      <p class="price">${{ product.price }}</p>
+      <p class="description">{{ product.description }}</p>
```
Comment (Kylyi) [inline components/products/Product.vue:19]: The design you chose is a bit more complex which is okay, but now there are quite ugly inconsistencies in between the `<Product />` rows - if one of the products has a long description for example, it stretches the row and it just looks bad.

image

Let's think about making it consistent.

File: components/products/Product.vue [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  padding-bottom: 10px;
+  width: 100%;
+}
+
+.image {
+  margin-right: 20px;
+  width: 50%;
+}
+
+.image-placeholder {
```
Comment (Kylyi) [inline components/products/Product.vue:37]: Could you explain this class please?

File: components/products/Product.vue [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+.price {
+  font-weight: bold;
+  margin: 10px 0;
+}
+
+.description {
+  margin-bottom: 0;
+}
+
+@media (max-width: 768px) {
```
Comment (Kylyi) [inline components/products/Product.vue:61]: I would prefer it to be mobile first, not the other way around

File: pages/product/new.vue [hunk 6]
```diff
@@ -0,0 +1,108 @@
+<script setup lang="ts">
+
+const product = reactive({
+    title: '',
+    description: '',
+    price: null,
+    image: '',
+});
+
+const source_image = ref('');
```
Comment (Kylyi) [inline pages/product/new.vue:10]: What is this for? :)

File: pages/product/new.vue [hunk 7]
```diff
@@ -0,0 +1,108 @@
+<script setup lang="ts">
+
+const product = reactive({
```
Comment (Kylyi) [inline pages/product/new.vue:3]: What is the difference between `ref` and `reactive`? And what is your opinion on the usage?

File: pages/product/new.vue [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    title: '',
+    description: '',
+    price: null,
```
Comment (Kylyi) [inline pages/product/new.vue:6]: You did not provide a type for the `product` variable, so the type here is inferred. In other words, now the typescript thinks that price is always a `null` which can be problematic for obvious reasons.

File: pages/product/new.vue [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    image: '',
+});
+
+const source_image = ref('');
+
+const submitForm = async () => {
```
Comment (Kylyi) [inline pages/product/new.vue:12]: I am missing at least some basic error handling.

File: pages/product/new.vue [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    const response = await $fetch('/api/products', {
+        method: 'POST',
+        body: JSON.stringify(product),
```
Comment (Kylyi) [inline pages/product/new.vue:15]: You don't need to stringify in nuxt :)

File: pages/product/new.vue
Comment (Kylyi) [inline pages/product/new.vue:1]: I really like the fact you chose to create a nested route but I would probably prefer if the route would be called `/productS/new` instead so it is consistent with the `/products` route

File: pages/product/new.vue [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        
+    if (file) {
+        const base64  = await convertToBase64(file);
+
+        // za boha nemůžu ten VueUse tady rozchodit, použil jsem normální reader a ten fungoval hned
+        //const { base64 } = useBase64(file)
+        //console.log(base64)
+
+        product.image = base64
+    }else product.image = ''
```
Comment (Kylyi) [inline pages/product/new.vue:31]: This is really ugly :)

File: pages/product/new.vue [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+// Function to handle file selection and convert image to BASE64
+async function handleImageSelect(event: InputEvent): Promise<void> {
+    const file = (event.target as HTMLInputElement).files?.[0];
+        
+    if (file) {
+        const base64  = await convertToBase64(file);
+
+        // za boha nemůžu ten VueUse tady rozchodit, použil jsem normální reader a ten fungoval hned
+        //const { base64 } = useBase64(file)
```
Comment (Kylyi) [inline pages/product/new.vue:27]: This is actually supposed to be used at the top level of the script, not inside the function.

File: pages/product/new.vue [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                <input type="text" id="title" name="title" placeholder="title" v-model="product.title" required>
+                
+                <label for="description">Description</label>
+                <textarea id="description" name="description" placeholder="description" v-model="product.description" required></textarea>
+                
+                <label for="price">Price</label>
+                <input type="number" step="0.01" id="price" name="price" placeholder="price" v-model="product.price" min="0" required>
+                
+                <label for="image">Image</label>
+                <input type="file" id="image" name="image" accept="image/png, image/jpeg, image/jpg" @change="handleImageSelect">
```
Comment (Kylyi) [inline pages/product/new.vue:64]: I really like the fact you are actually thinking about visually splitting the code so it's much more readable. But at some point, I would really like to see separation of the properties on different lines.

File: pages/product/new.vue [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    /* center content */
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+}
+form {
+    display: flex;
+    flex-direction: column;
+    gap: 10px;
+    width: 30%
```
Comment (Kylyi) [inline pages/product/new.vue:87]: Missing semicolon, but that's not really important. What is the more important thing here in terms of responsivity?

File: pages/product/new.vue
Comment (Kylyi) [inline pages/product/new.vue:1]: I've included the `@nuxthq/ui` UI library that was mentioned in the task's useful links part, it was also used in the `FileInput` component and is of course in the package.json. I'd assume you would use that for the form part but that's ok.

On the other hand, at least the component (and the only custom component) `FileInput` should have been definitely used.

Whenever you start working on a project that already has some base, you should use that base, not reinvent the wheel.

File: pages/products.vue [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+.product-grid {
+  display: grid;
+  grid-template-columns: repeat(2, minmax(250px, 1fr));
+  gap: 20px;
+}
+
+@media (max-width: 768px) {
+  .product-grid {
+    grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
```
Comment (Kylyi) [inline pages/products.vue:34]: Can you explain this line please?

File: server/api/products.post.ts [hunk 16]
```diff
@@ -1,9 +1,40 @@
-import { PrismaClient } from '@prisma/client'
+import { PrismaClient, Product } from '@prisma/client';
 
 const prisma = new PrismaClient()
 
 export default defineEventHandler(async (event) => {
   const body = await readBody(event)
 
-  // Implement your API here
+  if (!body.title || !body.price || !body.description) {
```
Comment (Kylyi) [inline server/api/products.post.ts:8]: I would like the error response to be more specific. Let's say I would like the error to say `Missing the product title. Missing the product price.` How would you do it?

File: pages/product/new.vue [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+const submitForm = async () => {
+    const response = await $fetch('/api/products', {
+        method: 'POST',
+        body: JSON.stringify(product),
+    })
+}
+
+// Function to handle file selection and convert image to BASE64
+async function handleImageSelect(event: InputEvent): Promise<void> {
```
Comment (Kylyi) [inline pages/product/new.vue:20]: You've used a different syntax to create the function here than in the section above.

What is the upside and downside of each usage? (using `const` to create a function OR using a `function`)

**Review by Kylyi — COMMENTED** [review]
In general, there are some inconsistencies but the code is on a decent level.

File: components/products/Product.vue (hunk 1 above)
Comment (flecmen) [reply components/products/Product.vue:13]: The missing img property might be a more specific "alt"?  Otherwise I'm not sure what other important property might be missing.

Malforming of the image could be fixed by setting a width in px. But then when using grid, the elements covered each other in some screen-widths. I fixed it then by using flex in products.

File: components/products/Product.vue (hunk 3 above)
Comment (flecmen) [reply components/products/Product.vue:19]: I fixed this with the previous issue, by setting width to both image and image-info. Plus using flex instead of grid in products

File: components/products/Product.vue (hunk 4 above)
Comment (flecmen) [reply components/products/Product.vue:37]: This class was meant to create a space for a missing image, when there is image missing in displayed product. To keep the structure symetric. It is definitely not ideal. It would be probably better to also include text saying " no image avaiable ", to explain the unused space to user.

File: pages/product/new.vue (hunk 14 above)
Comment (flecmen) [reply pages/product/new.vue:87]: Yes, I forgot to make this page responsive, so first thing would be making width=100% after some breakpoint. - Or the other way around, make it mobile-first and then shrink the width on larger screen. Also maybe remove padding on parent on mobile, to make it strech more to the sides.

File: pages/product/new.vue
Comment (flecmen) [reply pages/product/new.vue:1]: Oh, I missed that. Well it is sure better reusing already existing components and UI. I didn't really think about it here, hence I assumed it was a way to test my skills not using a UI library. Definitely my fault, made myself my work harder.

File: server/api/products.post.ts (hunk 16 above)
Comment (flecmen) [reply server/api/products.post.ts:8]: We could do 3 separate if statements, which I dont like because it repeats code and is hard to maintain if we would to alter Product attributes.
We could use validation library, like Zod, which is made for exactly this kind of checking. 
Or creating an array of required attributes and use that for checking the body, something like this:

const requiredFields = ['title', 'price', 'description'];
const missingFields = requiredFields.filter(field => !body[field]);

if (missingFields.length > 0) {
  const errorMessage = `Missing ${missingFields.join(', ')}`;
  return {
    statusCode: 400,
    body: errorMessage
  };
}

File: pages/product/new.vue (hunk 17 above)
Comment (flecmen) [reply pages/product/new.vue:20]: Difference is, that function declared with function can be called before declaration. With const, it has to be declared first. 

Also with function, "this" attribute represents object calling the function, and with const, "this" is same as in the code it is declared in.

File: components/products/Product.vue (hunk 1 above)
Comment (Kylyi) [reply components/products/Product.vue:13]: Hmm, setting a CSS `width` might not be the best idea when, for example, the image is of a lesser quality (let's say 200x100) and we would try to force the image to be of specific width (which might be bigger than the original image) - the quality of the image would be destroyed.

I was trying to lead you to setting an `object-fit: contain` (or smth similar) and more importantly setting `width` and `height` props for the img. You can probably ask chatgpt why that is quite important :)

File: components/products/Product.vue (hunk 2 above)
Comment (flecmen) [reply components/products/Product.vue:14]: Right, in this case, there would be "alt" text displayed. To prevent this, I would use @error tag and clean up the code, somewhat like this:

const imagePlaceholder = 'https://via.placeholder.com/600x400?text=Product+Image+Not+Found'

const handleImageError = (err: Event) => {
  (err.target as HTMLImageElement).src = imagePlaceholder;
}

Product Image

File: pages/products.vue (hunk 15 above)
Comment (flecmen) [reply pages/products.vue:34]: I see. The line 
`grid-template-columns: repeat(auto-fit, minmax(200px, 1fr)); `
doesnt appear to have much impact on the looks of the page. I inteded to make a responsive design, using the auto-fit. The real difference made the changed gap value.

File: components/products/Product.vue (hunk 2 above)
Comment (Kylyi) [reply components/products/Product.vue:14]: I was thinking about maybe conditionally putting `image-background` for the `` or something like that  - in general, i would try to go with the most css-like way.

File: pages/products.vue (hunk 15 above)
Comment (Kylyi) [reply pages/products.vue:34]: Honestly, I didnt really check if it does something, I was rather interested if you can explain this css to me in layman terms. In general, I like the idea

File: pages/product/new.vue
Comment (flecmen) [reply pages/product/new.vue:1]: > I really like the fact you chose to create a nested route but I would probably prefer if the route would be called `/productS/new` instead so it is consistent with the `/products` route

Yes, that is slightly inconsistent. I probably didn't like the fact, that the url /products/new suggests creating more products rather than only 1. Looking back, using /products seems more consistent and clear.

File: pages/product/new.vue (hunk 7 above)
Comment (flecmen) [reply pages/product/new.vue:3]: > What is the difference between `ref` and `reactive`? And what is your opinion on the usage?

ref and reactive are used to add reactivity to variables. **Ref** is used mainly for primitive values, **reactive** for objects. As I understand it, **ref** can be used for objects too, but the attributes loose reacitvity, while when using **reactive**, change of attribute triggers reactivity updates. 
**Ref** values are reached in the script using the .value, while **reactive** can be used as object directly.

Comment (flecmen) [reply pages/product/new.vue:3]: I prefer using refs as much as possible, but sometimes it is more convinient to wrap some complex data structure into reactive object.

File: pages/product/new.vue (hunk 8 above)
Comment (flecmen) [reply pages/product/new.vue:6]: > You did not provide a type for the `product` variable, so the type here is inferred. In other words, now the typescript thinks that price is always a `null` which can be problematic for obvious reasons.

Oops, that is true. But _product_ has some properties I don't need to declare when creating it, co I can solve this by using **Partial<>**:
`const product = reactive<Partial<Product>>({
    title: '',
    description: '',
    price: 0,
    image: '',
});`

File: pages/product/new.vue (hunk 6 above)
Comment (flecmen) [reply pages/product/new.vue:10]: > What is this for? :)

Probably leftovers of code from before I created reactive product object. This shouldn't be there of course.

File: pages/product/new.vue (hunk 9 above)
Comment (flecmen) [reply pages/product/new.vue:12]: > I am missing at least some basic error handling.

The form is checked on client side with **required** input attributes. But in the submit function, I could've added response code checking, and displaying some error message / success message.

File: pages/product/new.vue (hunk 12 above)
Comment (flecmen) [reply pages/product/new.vue:27]: > This is actually supposed to be used at the top level of the script, not inside the function.

But I'm getting the file from the event, where would I get it outside of the funciton? Or is it just the declaration?

File: pages/product/new.vue (hunk 11 above)
Comment (flecmen) [reply pages/product/new.vue:31]: > This is really ugly :)

I sometimes personally prefer not using {} if not needed for block of code or better clarity. But atleast I could've added some indentation:

}else 
        product.image = ''

If you find this not readible, I can just stick to curly braces every time.

File: pages/product/new.vue (hunk 13 above)
Comment (flecmen) [reply pages/product/new.vue:64]: > But at some point, I would really like to see separation of the properties on different lines.

This is something I usually leave on _prettier_ settings. I like properties on different lines as well.

File: pages/product/new.vue (hunk 11 above)
Comment (flecmen) [reply pages/product/new.vue:31]: *the code indentation isn't visible in the comment 😄.
